src/Text/Pretty/Simple/Internal/OutputPrinter.hs

```diff
@@ -176,82 +185,130 @@ hCheckTTY h options = liftIO $ conv <$> tty
-  sequenceFold
-    [ useColorNum
-    , pure (fromString number)
-    , useColorReset
-    ]
-renderOutput (Output _ (OutputStringLit string)) = do
-  options <- ask
-
-  sequenceFold
-    [ useColorQuote
-    , pure "\""
-    , useColorReset
-    , useColorString
-    -- TODO: This probably shouldn't be a string to begin with.
-    , pure (fromString (process options string))
-    , useColorReset
-    , useColorQuote
-    , pure "\""
-    , useColorReset
-    ]
+    subExpr x =
+      let doc = prettyExpr opts x
+      in
+        if isSimple x then
+          -- keep the expression on the current line
+          indent 1 doc
+        else
+          -- put the expression on a new line, indented
+          line <> indent (outputOptionsIndentAmount opts) doc
+
+-- | Construct a 'Doc' from a single 'Expr'.
+prettyExpr :: OutputOptions -> Expr -> Doc Annotation
+prettyExpr opts = \case
+  Brackets xss -> list "[" "]" xss
+  Braces xss -> list "{" "}" xss
+  Parens xss -> list "(" ")" xss
+  StringLit s -> join enclose (annotate Quote "\"") $ annotate String $ pretty s
+  CharLit s -> join enclose (annotate Quote "'") $ annotate String $ pretty s
+  Other s -> pretty s
+  NumberLit n -> annotate Num $ pretty n
   where
-    process :: OutputOptions -> String -> String
-    process opts = case outputOptionsStringStyle opts of
-      Literal -> id
-      EscapeNonPrintable ->
-        indentSubsequentLinesWith spaces . escapeNonPrintable . readStr
-      DoNotEscapeNonPrintable -> indentSubsequentLinesWith spaces . readStr
-      where
-        spaces :: String
-        spaces = replicate (indentSpaces + 2) ' '
-
-        indentSpaces :: Int
-        indentSpaces =  outputOptionsIndentAmount opts
-
-        readStr :: String -> String
-        readStr s = fromMaybe s . readMaybe $ '"':s ++ "\""
-renderOutput (Output _ (OutputCharLit string)) = do
-  sequenceFold
-    [ useColorQuote
-    , pure "'"
-    , useColorReset
-    , useColorString
-    , pure (fromString string)
-    , useColorReset
-    , useColorQuote
-    , pure "'"
-    , useColorReset
-    ]
+    list :: Doc Annotation -> Doc Annotation -> CommaSeparated [Expr]
+      -> Doc Annotation
+    list open close (CommaSeparated xss) =
+      enclose (annotate Open open) (annotate Close close) $ case xss of
+        [] -> mempty
+        [xs] | all isSimple xs ->
+          space <> hsep (map (prettyExpr opts) xs) <> space
+        _ -> concatWith lineAndCommaSep (map (prettyExprs opts) xss) <> line
+    lineAndCommaSep x y = x <> line <> annotate Comma "," <> y
+
+-- | Determine whether this expression should be displayed on a single line.
+isSimple :: Expr -> Bool
+isSimple = \case
+  (getList -> Just [[e]]) -> isSimple e
+  (getList -> Just (_:_)) -> False
+  _ -> True
+  where
+    getList = \case
+      Brackets (CommaSeparated xs) -> Just xs
+      Braces (CommaSeparated xs) -> Just xs
+      Parens (CommaSeparated xs) -> Just xs
+      _ -> Nothing
```

Comment:
Wouldn't it be simpler to inline `getList` into the `case` expression above?

Reply:
Just inlining it would result in some slightly repetitive code, but this does seem a strange approach now (an earlier version was more complicated). Good catch. Changed in f6ae384e.